warp10/src/main/java/io/warp10/script/aggregator/Median.java

```diff
@@ -26,147 +26,107 @@
-    for (int i = 0; i < values.length; i++) {
-      if (null != values[i]) {
-        if (values[i] instanceof Number) {
-          if (!(values[i] instanceof Double && Double.isNaN((Double) values[i]))) {
-            validValues.add((Number) values[i]);
-          }
+    Arrays.sort(indices, new Comparator<Integer>() {
+      @Override
+      public int compare(Integer idx1, Integer idx2) {
+        if (null == values[idx1] && null == values[idx2]) {
+          return 0;
+        } else if (null == values[idx1] || null == values[idx2]) {
+          return null == values[idx1] ? 1 : -1;
+        } else if (values[idx1] instanceof Number && values[idx2] instanceof Number) {
+          return EQ.compare((Number) values[idx1], (Number) values[idx2]);
         } else {
-          throw new WarpScriptException(this.getName() + " cannot compute median of non numeric values.");
+          throw new RuntimeException("MEDIAN can only operate on numeric Geo Time Series.");
         }
       }
-    }
-
-    //
-    // Sort values
-    // Could be Long or Double, or a mix, if used as a reducer.
-    //
-    Collections.sort(validValues, new Comparator<Number>() {
-      @Override
-      public int compare(Number o1, Number o2) {
-        return EQ.compare(o1, o2);
-      }
     });
 
+    long location = 0;
+    long elevation = GeoTimeSerie.NO_ELEVATION;
+    Object median;
 
-    Object median = null;
+    int nonNullLength = values.length - nullCounter;
 
-    if (0 != validValues.size()) {
-      if (validValues.get(0).equals(validValues.get(validValues.size() - 1))) {
-        // If extrema are identical, use this as the median
-        median = validValues.get(0);
-      } else {
-        int len = validValues.size();
-        if (0 == len % 2) {
-          Object low = validValues.get((len / 2) - 1);
-          Object high = validValues.get(len / 2);
-          if (low instanceof Long && high instanceof Long) {
-            median = ((long) low + (long) high) / 2L;
-          } else {
-            median = ((double) low + (double) high) / 2.0D;
-          }
-        } else {
-          median = validValues.get(len / 2);
+    //
+    // singleton case
+    //
+    if (1 == nonNullLength) {
+      return new Object[]{
+          tick, locations[indices[0]], elevations[indices[0]], values[indices[0]]
+      };
+    } else {
+      if (0 == nonNullLength % 2) {
+        //
+        // even number of non null values, return mean of both values.
+        // If there is a location for both points, return centroid of locations
+        // If there is an elevation for both points, return mean of elevations
+        //
+        int low = indices[nonNullLength / 2 - 1];
+        int high = indices[nonNullLength / 2];
+        median = (((Number) values[low]).doubleValue() + ((Number) values[high]).doubleValue()) / 2.0D;
+        if (GeoTimeSerie.NO_ELEVATION != elevations[low] && GeoTimeSerie.NO_ELEVATION != elevations[high]) {
```

Comment:
If there is only one location (same goes for elevation), wouldn't it be better to return that location (resp. elevation)?

Reply:
To be discussed... What is your preference ?

Reply:
fixed.